### ISSUE TITLE
role="status" on ContextSuccessMessage and ContextErrorMessage

### ISSUE THREAD

**Issue by torkjels** [SpareBank1/designsystem#795]
#### Which package is this issue related to?
ffe-context-message-react

#### Describe your issue (screenshots welcome!)
During an accessibility workshop today by Morten Tollefsen from MediaLT, it was brought up that screen readers are not notified of dynamically added `ContextMessage` elements. As a result, users using technologies such as screen readers tend to miss these error and success messages.

#### What is the expected behavior? 
The message is read out by the screen reader

#### What is the actual behavior?
The message is not read out by the screen reader. Users with visual impairments won't be notified of success/error messages.

#### Steps to reproduce
In the browser, make a transfer between two of your accounts and there will pop up a green `ContextSuccessMessage` telling you that the transfer will be completed on your due date. However, this information isn't presented to people using screen readers as there is no `role="alert"` or `aria-live` attribute on the `ContextMessage` elements.

**Comment by ivarni** [comment]
This can easily be solved at the application-level by providing the `role` prop when needed since we're spreading any unknown props directly to the root element of the component (https://github.com/SpareBank1/designsystem/blob/develop/packages/ffe-context-message-react/src/ContextMessage.js#L76,L85)

**Comment by ivarni** [comment]
For messages that are not added dynamically setting this prop would be wrong as far as I know, and the component itself has no way of knowing so this seems like something that should be handled outside of it.

**Comment by torkjels** [comment]
The issue with passing it as a prop is that I imagine "no one" will do it, which is how we end up where we are today. That's why I think we should put as much accessibility responsibility as possible on the design system itself.

Maybe my imagination is a bit lackluster, but are there really any scenarios where you wouldn't want an alert on at least the success and error messages? Maybe in the instances where you don't want an alert (which is only a few - from my understanding) we could rather specifically override the role, telling the message *not* to alert.

**Comment by ivarni** [comment]
From the [mdn page](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_alert_role)

> Because of its intrusive nature, the alert role must be used sparingly and only in situations where the user's immediate attention is required. Dynamic changes that are less urgent should use a less aggressive method, such as aria-live="polite" or other live region roles.

It doesn't seem like a good idea to me to leave that as a default. 

I think a much better approach is to raise awareness amongst developers so that they can make informed and deliberate decisions about when to apply ARIA roles. Swinging the pendulum too far in the other direction won't necessarily be an improvement in my opinion.

I'm not sure a success message warrants requiring a user's "immediate attention" but an error message might. It kind of depends. A scenario where I imagine it's not needed is if we're accepting a form submission and then sending the user to a receipt page where there's an error message. In that case it's not added dynamically and I'm not sure the alert role would be needed since the entire DOM was just changed and needs to be announced anyway?

For the [form errror message](https://github.com/SpareBank1/designsystem/blob/develop/packages/ffe-form-react/src/InputGroup.js#L106) we probably should use it, because at least there the context we're showing the message in is already known to the component.

**Comment by Snikanes** [comment]
During the workshop, it was made clear that Morten Tollefsen could not tell _at all_ whether his action had succeeded or failed. To those with normal eyesight these context messages convey all their information at a single glance, while Morten did not receive a single piece of information regarding the outcome of his action. I think both success messages and error messages both warrant some kind of notification to the screen reader by way of ARIA roles.

I disagree that raising awareness is necessarily the best way to solve this. This will inevitably lead to some teams adding these ARIA roles and other teams refraining from doing so which will lead to inconsistencies in when to expect these messages. 

I do agree, however, that `role=alert` might not be the best choice here. @ivarni your quote mentions an alternative `aria-live="polite"` that is less intrusive. Another alternative that might suit our needs is the `role=status`, which is an extension to `aria-live="polite"`: https://www.w3.org/WAI/WCAG21/Techniques/aria/ARIA22.html

**Comment by ivarni** [comment]
I'm not arguing that the application under test wasn't malfunctioning, I'm worried about adding too eager defaults to the visual components in the design system, that's all.

**Comment by Snikanes** [comment]
I was using the application to state an example. Aren't these messages meant to be used as a response to a user action? To normal-sighted users that is exactly what they accomplish. To anyone using a screen reader, these messages are not distinguished from other elements on the page. As such, they do not do what they are supposed to do.

> I'm not sure a success message warrants requiring a user's "immediate attention" but an error message might. It kind of depends. A scenario where I imagine it's not needed is if we're accepting a form submission and then sending the user to a receipt page where there's an error message. In that case it's not added dynamically and I'm not sure the alert role would be needed since the entire DOM was just changed and needs to be announced anyway?

Why would we want to distinguish between whether the success/error message was added dynamically or present on initial render?

**Comment by ivarni** [comment]
I don't know why our designers use them as they do but I've very rarely, if ever, seen them being used in a manner that would demand immediate attention.

My only point is that eagerly applying defaults like this can end up being just as bad for the end user as not training developers in how to use components in an accessible way.

 > Why would we want to distinguish between whether the success/error message was added dynamically or present on initial render?

If it's already the first element rendered on a fresh page, why would it need a role to further emphasise its importance? It's going to get announced first anyway, isn't it?

**Comment by torkjels** [comment]
> I don't know why our designers use them as they do but I've very rarely, if ever, seen them being used in a manner that would demand immediate attention.

We might define "demand immediate attention" differently. But then again, the "immediate attention" is mostly a concern when we're talking about `role="alert"`, but I agree with both of you that `role="alert"` might be a bit too much. So as @Snikanes suggests, `role="status"` would make more sense.
Maybe the success/error message don't require *immediate* attention, but I'd argue that they do indeed require attention (although not *immediate*). A `role="status"` (or some other form of `aria-live="polite"`) would accomplish that.
I'd argue that these success/error messages require attention from users. They tell you if your action was successful or not, or if something generally is terribly wrong. Their purpose is to show this to the user in a way that demands attention - even if that attention is just a quick read. This should apply to people using screen readers too.

> If it's already the first element rendered on a fresh page, why would it need a role to further emphasise its importance? It's going to get announced first anyway, isn't it?

If not for any emphasis, it's only gonna be announced when the user focuses it.
When sent to a new page, the focus is first put at the top of the page again, all the way up in the header, losing the context of what they were doing and having no idea if their action was successful or not. The only way for them to know if their action was successful is to then navigate through the page in hopes that they'll see some success/error message somewhere.

And in some cases the success/error message might not be at the top of the page. Maybe someone puts their error message close to the submit button all the way at the bottom of the page.

**Comment by ivarni** [comment]
Fair enough, and good points :)

I'd be in favour of adding the `status` role as a default (or as a an alternative adding `role` as a required prop but I don't know what the culture is for keeping a clean console is in different teams so it might be better to just roll a default).

**Comment by torkjels** [comment]
Alright! You bring up some good points too, though. We don't want to go overboard. I'll contact the accessibility expert that held the workshop for us yesterday and ask him if he thinks this specific solution would be overkill or not.

**Comment by pethel** [comment]
Interesting discussion. The major issue is that we are doing a lot of asynchronous actions, while fetching and storing data.  I was looking into this(aria-live) generally a bit before an stumbled into this one. 

https://almerosteyn.com/2017/09/aria-live-regions-in-react

The intresting part:

> One of the most important things to keep in mind when making use of an ARIA live region is that changes to the content will only be read out by screen readers if the entire region already exists in the DOM and there has been any changes.
> 
> This means we cannot render the live region in the DOM for the first time at the moment we want to dispatch a message to the screen reader. No, on the initial render the inside text is typically ignored by screen readers.
> 
> And here comes the caveat in a React application; we need to ensure that the live region is already rendered when we send our first message and that it stays rendered until we no longer require it.
> 

This also suggests that the contet as must be present in the DOM on page load.
https://www.w3.org/TR/WCAG20-TECHS/ARIA19.html

I haven't actually tried this on a screen reader but my point is that I think that setting aria-live on a new element is it not enough.  
Maybe something like this, [react-aria-live](https://github.com/AlmeroSteyn/react-aria-live) or your own solution.

**Comment by torkjels** [comment]
That's a great point, @pethel. I tested this out using [ChromeVox](https://chrome.google.com/webstore/detail/chromevox-classic-extensi/kgejglhpjiefppelpmljglcjbhoiplfn), but it might work completely differently in other more "full-fledged" screen readers.
At least this example works with ChromeVox: https://codepen.io/tskjetne/pen/QWwddqj

**Comment by pethel** [comment]
I do not know, maybe things have changed? I never tested this myself.
I guess the best thing we can do is test this on the most popular screen readers.

**Comment by pethel** [comment]
Lenge seden sist her men i følge kurserna til https://dequeuniversity.com/ så ær ikke ChromeVox representativ og noe man burde bruke for og finne ut av dette.

**Comment by kwltrs** [comment]
ContextMessages brukes på mange forskjellige måte, så vi kan ikke sette role-attrbutet per default og må overlate ansvaret til appen.
